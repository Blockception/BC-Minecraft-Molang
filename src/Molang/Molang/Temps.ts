import { RegularExpression } from "../RegExp";

/**The namespace that governs molang temp variables*/
export namespace Temps {
  /**The pattern used to find the defintions of temp variables*/
  export const getDefinedPatt: RegExp = /(?:^|;[ \t]*|"[ \t]*)\b(?:t|temp)\.([a-z0-9_]*)[ \t]*\=/gim;

  /**The pattern used to find the temp variables*/
  export const getUsedPatt: RegExp = /\b(?:t|temp)\.([a-z0-9_]+)\b(?![ \t]+=)/gim;

  /**Spits through all the provided strings searching for all instances of defining temp variables.
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getDefined(data: any, receiver: string[]): void {
    RegularExpression.harvest(data, getDefinedPatt, receiver);
  }

  /**Spits through all the provided strings searching for all instances of using temp variables.
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getUsing(data: any, receiver: string[]): void {
    RegularExpression.harvest(data, getUsedPatt, receiver);
  }
}
