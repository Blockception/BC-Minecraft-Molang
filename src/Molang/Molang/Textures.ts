import { RegularExpression } from "../RegExp";

/**The namespace that governs molang textures*/
export namespace Textures {
  /**The pattern used to find the used textures*/
  export const getUsedPatt: RegExp = /\b(?:texture)\.([a-z0-9_]+)/gim;

  /**Spits through all the provided strings searching for all instances of using textures
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getUsing(data: any, receiver: string[]): void {
    RegularExpression.harvest(data, getUsedPatt, receiver);
  }

  /**The namespace that governs molang arrays*/
  export namespace Arrays {
    /**The pattern used to find the used arrays*/
    export const getUsedPatt: RegExp = /\b(?:array)\.([a-z0-9_]+)/gim;

    /**Spits through all the provided strings searching for all instances of using arrays
     * @param data The string(s)/container to look through
     * @param receiver The receiving array*/
    export function getUsing(data: any, receiver: string[]): void {
      RegularExpression.harvest(data, getUsedPatt, receiver);
    }
  }
}
