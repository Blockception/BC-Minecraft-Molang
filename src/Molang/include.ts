/*	Auto generated	*/

export * as Molang from "./Molang/include";
export * from "./MolangSet";
export * from "./RegExp";
