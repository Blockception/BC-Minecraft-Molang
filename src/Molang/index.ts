/*	Auto generated	*/
export * as Types from "./Types/index";
export * from "./Functions";
export * from "./MolangSet";
export * from "./RegExp";
export * from "./Syntax";
export * from "./Traverese";