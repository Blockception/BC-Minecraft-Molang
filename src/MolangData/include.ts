import { Animations } from "./Animations";
import { AnimationsControllers } from "./AnimationsControllers";
import { Blocks } from "./Blocks";
import { Entities } from "./Entities";
import { FeaturesRules } from "./FeaturesRules";
import { Items } from "./Items";
import { Particles } from "./Particles";
import { RenderControllers } from "./RenderControllers";

export * from "./Data";
export * from "./General";
export { Animations, AnimationsControllers, Blocks, Entities, FeaturesRules, Items, Particles, RenderControllers };

/**Returns either an object with molang data or undefined
 * @param id The identification of molang data to return
 * @returns A namespace carrying the data or undefined if nothing was found*/
export function get(
  id: "animation" | "animation_controller" | "block" | "entity" | "feature" | "item" | "particle" | "render_controller"
): typeof Blocks | undefined {
  switch (id) {
    case "animation":
      return Animations;

    case "animation_controller":
      return AnimationsControllers;

    case "block":
      return Blocks;

    case "entity":
      return Entities;

    case "feature":
      return FeaturesRules;

    case "item":
      return Items;

    case "particle":
      return Particles;

    case "render_controller":
      return RenderControllers;
  }

  return undefined;
}
