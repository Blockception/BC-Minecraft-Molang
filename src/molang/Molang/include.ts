/*	Auto generated	*/

export * from "./Context";
export * from "./Geometries";
export * from "./Materials";
export * from "./Queries";
export * from "./Temps";
export * from "./Textures";
export * from "./Variables";
