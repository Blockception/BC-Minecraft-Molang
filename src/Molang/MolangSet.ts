import { DefinedUsing } from "../Types/Defined Using/DefinedUsing";
import { Using } from "../Types/Defined Using/include";
import { Context, Queries, Temps, Variables, Geometries, Textures, Materials } from "./Data/include";

/** */
export interface MolangSet {
  /** */
  contexts: Using<string>;
  /** */
  queries: Using<string>;
  /** */
  variables: DefinedUsing<string>;
  /** */
  temps: DefinedUsing<string>;
}

/** */
export interface MolangFullSet extends MolangSet {
  /** */
  materials: DefinedUsing<string>;
  /** */
  textures: DefinedUsing<string>;
  /** */
  geometries: DefinedUsing<string>;
}

/** */
export namespace MolangSet {
  /**
   *
   * @returns
   */
  export function create(): MolangSet {
    return {
      queries: Using.create<string>(),
      variables: DefinedUsing.create<string>(),
      temps: DefinedUsing.create<string>(),
      contexts: Using.create<string>(),
    };
  }

  /**
   *
   * @param object
   * @returns
   */
  export function harvest(object: any): MolangSet {
    const out = create();

    Queries.getUsing(object, out.queries.using);
    Context.getUsing(object, out.contexts.using);

    Variables.getUsing(object, out.variables.using);
    Variables.getDefined(object, out.variables.defined);

    Temps.getUsing(object, out.temps.using);
    Temps.getDefined(object, out.temps.defined);

    return out;
  }
}

/**
 *
 */
export namespace MolangFullSet {
  /**
   *
   * @returns
   */
  export function create(): MolangFullSet {
    return {
      queries: Using.create<string>(),
      variables: DefinedUsing.create<string>(),
      temps: DefinedUsing.create<string>(),
      contexts: Using.create<string>(),
      geometries: DefinedUsing.empty(),
      materials: DefinedUsing.empty(),
      textures: DefinedUsing.empty(),
    };
  }

  /**Add the necessary properties to the given data set to the full MolangFullSet
   * @param data
   */
  export function upgrade(data: MolangSet): MolangFullSet {
    const out = <MolangFullSet>data;

    if (!DefinedUsing.is<string>(out.geometries)) out.geometries = DefinedUsing.empty();
    if (!DefinedUsing.is<string>(out.materials)) out.materials = DefinedUsing.empty();
    if (!DefinedUsing.is<string>(out.textures)) out.textures = DefinedUsing.empty();

    return out;
  }

  /**
   *
   * @param object
   */
  export function harvest(object: object | string): MolangFullSet {
    const out = create();

    Queries.getUsing(object, out.queries.using);
    Context.getUsing(object, out.contexts.using);

    Variables.getUsing(object, out.variables.using);
    Variables.getDefined(object, out.variables.defined);

    Temps.getUsing(object, out.temps.using);
    Temps.getDefined(object, out.temps.defined);

    Textures.getUsing(object, out.textures.using);
    Geometries.getUsing(object, out.geometries.using);
    Materials.getUsing(object, out.materials.using);

    return out;
  }

  /**
   *
   * @param value
   */
  export function isEither(value: MolangSet | MolangFullSet): value is MolangFullSet {
    const temp = <MolangFullSet>value;

    if (typeof temp.geometries === "object" && typeof temp.materials === "object" && typeof temp.textures === "object") {
      return true;
    }

    return false;
  }
}
