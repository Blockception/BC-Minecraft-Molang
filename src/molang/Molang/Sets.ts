import { RegularExpression } from "../RegExp";
import { MolangFullSet, MolangSet } from "../MolangSet";

/**The namespace that governs molang sets*/
export namespace Sets {
  /**The pattern used to find the used to find most items*/
  export const getUsedPatt: RegExp = /\b(c|context|t|temp|q|query|m|math|v|variable)\.([a-z0-9_]+)\b(?![ \t]+=)/gim;
  /**The pattern used to find the used to find all items*/
  export const getFullUsedPatt: RegExp =
    /\b((t|temp|v|variable)\.([a-z0-9_]+)\b(?![ \t]+=)|(c|context|q|query|m|math|texture|geometry|material|array)\.([a-z0-9_][a-z0-9_\.]+)\b)/gim;
  /**The pattern used to find the used to find all defined items*/
  export const getDefinedPatt: RegExp = /(?:^|;[ \t]*|"[ \t]*)\b(v|t|variable|temp)\.([a-z0-9_]*)[ \t]*\=/gim;

  /**Spits through all the provided strings searching for all instances of using textures
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getUsingFull(data: any, set: MolangFullSet): void {
    const receiver: string[] = [];
    RegularExpression.harvest(data, getFullUsedPatt, receiver);

    console.log(JSON.stringify(receiver));

    for (let I = 0; I < receiver.length; I++) {
      const type = receiver[I++];
      const value = receiver[I];

      if (type && value) {
        switch (type.toLowerCase()) {
          case "c":
          case "context":
            set.contexts.using.push(value);
            break;

          case "v":
          case "variable":
            set.variables.using.push(value);
            break;

          case "t":
          case "temp":
            set.temps.using.push(value);
            break;

          case "q":
          case "query":
            set.queries.using.push(value);
            break;

          case "math":
            //TODO
            break;

          case "texture":
            set.textures.using.push(value);
            break;

          case "geometry":
            set.geometries.using.push(value);
            break;

          case "material":
            set.materials.using.push(value);
            break;
        }
      }
    }
  }

  /**Spits through all the provided strings searching for all instances of using textures
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getUsing(data: any, set: MolangSet): void {
    const receiver: string[] = [];
    RegularExpression.harvest(data, getUsedPatt, receiver);

    for (let I = 0; I < receiver.length; I++) {
      const type = receiver[I++];
      const value = receiver[I];

      if (type && value) {
        switch (type.toLowerCase()) {
          case "c":
          case "context":
            set.contexts.using.push(value);
            break;

          case "v":
          case "variable":
            set.variables.using.push(value);
            break;

          case "t":
          case "temp":
            set.temps.using.push(value);
            break;

          case "q":
          case "query":
            set.queries.using.push(value);
            break;

          case "math":
            //TODO
            break;
        }
      }
    }
  }

  /**Spits through all the provided strings searching for all instances of using textures
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getDefined(data: any, set: MolangSet): void {
    const receiver: string[] = [];
    RegularExpression.harvest(data, getDefinedPatt, receiver);

    for (let I = 0; I < receiver.length; I++) {
      const type = receiver[I++];
      const value = receiver[I];

      if (type && value) {
        switch (type.toLowerCase()) {
          case "v":
          case "variable":
            set.variables.defined.push(value);
            break;

          case "t":
          case "temp":
            set.temps.defined.push(value);
            break;
        }
      }
    }
  }
}
