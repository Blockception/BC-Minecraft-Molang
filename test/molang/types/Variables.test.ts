import { expect } from "chai";
import { Types } from '../../../src/Molang/include';
import { VanillaPlayer } from "../../Player.test";

describe("molang", () => {
  describe("variables", () => {
    it("defined1", () => {
      let receiver: string[] = [];

      Types.Variables.getDefined("variable.foo1 = 0; variable.foo2 = 0; v.foo3 = 0;", receiver);

      expect(receiver).to.have.members(["foo1", "foo2", "foo3"]);
    });

    it("defined2", () => {
      let receiver: string[] = [];

      Types.Variables.getDefined(VanillaPlayer.DataOBject, receiver);

      expect(receiver).to.contain.members(VanillaPlayer.Variables.defined);
    });

    it("defined3", () => {
      let receiver: string[] = [];

      Types.Variables.getDefined(VanillaPlayer.Data, receiver);

      expect(receiver).to.contain.members(VanillaPlayer.Variables.defined);
    });

    it("defined special",()=>{
      const data = `"initialize": [
        "variable.armor_stand.pose_index = 0;",
        "variable.armor_stand.hurt_time = 0;"
      ]`;

      let receiver : string[] = [];
      Types.Variables.getDefined(data, receiver);

      expect(receiver).to.contains.members(["armor_stand.pose_index", "armor_stand.hurt_time"])
    })

    it("duplicate check", () => {
      let receiver: string[] = [];

      Types.Variables.getDefined(["variable.foo1 = 0; variable.foo2 = 0; v.foo3 = 0;", "variable.foo3 = 0;"], receiver);

      expect(receiver).to.have.members(["foo1", "foo2", "foo3"]);
    });

    it("defined not sticky?", () => {
      expect(Types.Variables.getDefinedPatt.sticky).to.equal(false);
    });

    it("using1", () => {
      let receiver: string[] = [];

      Types.Variables.getUsing(VanillaPlayer.Data, receiver);

      expect(receiver).to.contain.members(VanillaPlayer.Variables.using);
    });

    it("using2", () => {
      let receiver: string[] = [];

      Types.Variables.getUsing(VanillaPlayer.DataOBject, receiver);

      expect(receiver).to.contain.members(VanillaPlayer.Variables.using);
    });
  });
});
