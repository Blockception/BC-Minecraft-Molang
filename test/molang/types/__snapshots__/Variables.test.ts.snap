// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molang variables defined special 1`] = `
[
  "armor_stand.pose_index",
  "armor_stand.hurt_time",
]
`;

exports[`Molang variables defined1 1`] = `
[
  "foo1",
  "foo2",
  "foo3",
]
`;

exports[`Molang variables defined2 1`] = `
[
  "is_holding_right",
  "is_blinking",
  "last_blink_time",
  "hand_bob",
  "helmet_layer_visible",
  "leg_layer_visible",
  "boot_layer_visible",
  "chest_layer_visible",
  "attack_body_rot_y",
  "tcos0",
  "first_person_rotation_factor",
  "map_angle",
  "item_use_normalized",
]
`;

exports[`Molang variables defined3 1`] = `
[
  "is_holding_right",
  "is_blinking",
  "last_blink_time",
  "hand_bob",
  "helmet_layer_visible",
  "leg_layer_visible",
  "boot_layer_visible",
  "chest_layer_visible",
  "attack_body_rot_y",
  "tcos0",
  "first_person_rotation_factor",
  "map_angle",
  "item_use_normalized",
]
`;

exports[`Molang variables duplicate check 1`] = `
[
  "foo1",
  "foo2",
  "foo3",
]
`;

exports[`Molang variables using1 1`] = `
[
  "is_holding_right",
  "is_blinking",
  "last_blink_time",
  "hand_bob",
  "helmet_layer_visible",
  "leg_layer_visible",
  "boot_layer_visible",
  "chest_layer_visible",
  "attack_body_rot_y",
  "attack_time",
  "tcos0",
  "gliding_speed_value",
  "first_person_rotation_factor",
  "map_angle",
  "player_x_rotation",
  "item_use_normalized",
  "is_first_person",
  "map_face_icon",
]
`;

exports[`Molang variables using2 1`] = `
[
  "is_holding_right",
  "is_blinking",
  "last_blink_time",
  "hand_bob",
  "helmet_layer_visible",
  "leg_layer_visible",
  "boot_layer_visible",
  "chest_layer_visible",
  "attack_body_rot_y",
  "attack_time",
  "tcos0",
  "gliding_speed_value",
  "first_person_rotation_factor",
  "map_angle",
  "player_x_rotation",
  "item_use_normalized",
  "is_first_person",
  "map_face_icon",
]
`;

exports[`Molang variables using3 1`] = `
[
  "armor_stand.pose_index",
]
`;
